services/code_hosting/git_hub_service.rb

```diff
@@ -1,131 +1,409 @@
 require_relative "code_hosting_service"
 require_relative "../../shared/logging_module"
+require_relative "../../shared/models/project"
+require_relative "../../fastfile-parser/fastfile_parser"
 
 require "set"
 require "octokit"
+require "git"
+require "addressable/uri"
+require "tty-command"
+require "securerandom"
+require "digest"
+require "pathname"
 
 module FastlaneCI
   # Data source that interacts with GitHub
   class GitHubService < CodeHostingService
     include FastlaneCI::Logging
 
     class << self
-      attr_accessor :status_context_prefix
-    end
+      attr_writer :temporary_git_storage
+
+      attr_accessor :temporary_storage_path
+
+      attr_writer :cache
+
+      attr_accessor :git_action_queue
+
+      def cache
+        @cache ||= {}
+        return @cache
+      end
+
+      # Client for GitHub-related operations, uses an in-memory cache
+      # to fast things.
+      # @param [String] api_token
+      def client(api_token)
```

Comment:
I think this might be premature optimization. I don't think the Octokit::Client is expensive to build. It might also be best recreate them anyway.

Reply:
Replied below |
                        V

Reply:
My main concern is that there's a lot of operations that relies on the `Octokit::Client` inside this class, so memoize it for a given `api_token` seem kinda nice, and allows future code hosting services to adopt this approach easily (as it's implemented yet)

Reply:
A couple things: I'm not familiar enough with `Octokit::Client` to know if it has an age limit on its own authentication, or if it manages its own internal state perfectly. Often frameworks do some sort of handshake and token exchange, and if you let a client sit for too long, the current auth gets stale and then reauth is needed, which often isn't implemented correctly. Also, the longer an object lives, the more opportunities for state to change and break things. So I prefer to just create new objects until performance testing shows that we need caching.

By not using caches until its obvious that caches are needed means that you eliminate the possibility of running into many classes of bugs hard to track down problems in other people's code. Also, caching always seems deceptively easy until it's not. In my experience it's one of the largest sources of problems.

As a rule of thumb, it's best to avoid caches until you have proof that they are needed, and that they actually fix a problem.